pkg/cmd/operator/cmd.go

```diff
@@ -67,35 +76,95 @@ func NewOperator() *cobra.Command {
 				log.WithError(err).Fatal("unable to set up client config")
 			}
 
-			// Create a new Cmd to provide shared dependencies and start components
-			log.Info("setting up manager")
-			mgr, err := manager.New(cfg, manager.Options{
-				MetricsBindAddress:      ":2112",
-				LeaderElection:          true,
-				LeaderElectionNamespace: minterv1.CloudCredOperatorNamespace,
-				LeaderElectionID:        leaderElectionConfigMap,
-			})
-			if err != nil {
-				log.WithError(err).Fatal("unable to set up overall controller manager")
+			run := func(ctx context.Context) {
+				// Create a new Cmd to provide shared dependencies and start components
+				log.Info("setting up manager")
+				mgr, err := manager.New(cfg, manager.Options{
+					MetricsBindAddress: ":2112",
+				})
+				if err != nil {
+					log.WithError(err).Fatal("unable to set up overall controller manager")
+				}
+
+				log.Info("registering components")
+
+				// Setup Scheme for all resources
+				util.SetupScheme(mgr.GetScheme())
+
+				// Setup all Controllers
+				log.Info("setting up controller")
+				kubeconfigCommandLinePath := cmd.PersistentFlags().Lookup("kubeconfig").Value.String()
+				if err := controller.AddToManager(mgr, kubeconfigCommandLinePath); err != nil {
+					log.WithError(err).Fatal("unable to register controllers to the manager")
+				}
+
+				// Start the Cmd
+				log.Info("starting the cmd")
+				if err := mgr.Start(signals.SetupSignalHandler()); err != nil {
+					log.WithError(err).Fatal("unable to run the manager")
+				}
 			}
 
-			log.Info("registering components")
-
-			// Setup Scheme for all resources
-			util.SetupScheme(mgr.GetScheme())
-
-			// Setup all Controllers
-			log.Info("setting up controller")
-			kubeconfigCommandLinePath := cmd.PersistentFlags().Lookup("kubeconfig").Value.String()
-			if err := controller.AddToManager(mgr, kubeconfigCommandLinePath); err != nil {
-				log.WithError(err).Fatal("unable to register controllers to the manager")
+			// Leader election code based on:
+			// https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/examples/leader-election/main.go#L130
+			// This gives us ReleaseOnCancel which is not presently exposed in controller-runtime.
+
+			// use a Go context so we can tell the leaderelection code when we want to step down
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			// listen for interrupts or the Linux SIGTERM signal and cancel
+			// our context, which the leader election code will observe and
+			// step down
+			ch := make(chan os.Signal, 1)
+			signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+			go func() {
+				<-ch
+				log.Info("received termination, signaling shutdown")
+				cancel()
+			}()
+
+			id := uuid.New().String()
+			leLog := log.WithField("id", id)
+			leLog.Info("generated leader election ID")
+
+			lock := &resourcelock.ConfigMapLock{
```

Comment:
Is this a good time to switch to using a `LeaseLock` instead of a `ConfigMapLock`? Or is this more complicated due to needing to upgrade existing CCOs that would still be using a `ConfigMapLock`?

Reply:
I wasn't totally aware how leaselock works, doesn't appear to be a resource in-cluster at least in 4.5 where I was running this. There is potential for some problems during upgrade though for sure, I'd vote to stay consistent for now.